File: api/app/modules/lp/graph_nd.py
"""Dispatch the LP picture: plane (2 vars), polyhedron (3), or a slice."""

from __future__ import annotations

from app.modules.lp.graph_2d import _clean, _fmt_num, build_2d_graph
from app.modules.lp.graph_3d import build_3d_graph
from app.modules.lp.models import LPConstraint, LPRequest
from app.schemas.result import GraphXY

EPS = 1e-9


def build_lp_graph(
    req: LPRequest, x_star: dict[str, float], z_star: float
) -> tuple[GraphXY | None, list[str]]:
    names = _decision_names(req)
    warnings: list[str] = []
    if len(names) < 2:
        return None, warnings

    free, unknown = _free_names(req, names)
    if unknown:
        warnings.append(
            "Variables de gráfico ignoradas porque no están en el modelo: " + ", ".join(unknown)
        )
    if len(free) < 2:
        return None, warnings

    sliced = set(free) != set(names)
    if not sliced and len(free) == 2:
        return build_2d_graph(req, x_star, z_star), warnings

    reduced, z_offset, fixed = _reduce(req, x_star, free)
    z_free = z_star - z_offset
    if len(free) == 2:
        graph = build_2d_graph(reduced, x_star, z_free, z_offset)
    else:
        graph = build_3d_graph(reduced, x_star, z_free, z_offset, fixed if sliced else None)
        if graph is None and sliced:
            warnings.append(
                "No se pudo construir el poliedro 3D. Elige otras tres variables o un corte de dos."
            )
    if graph is None:
        return None, warnings
    if sliced and len(free) == 2:
        held = ", ".join(f"{name} = {_fmt_num(value)}" for name, value in fixed.items())
        graph = graph.model_copy(
            update={
                "title": "Corte por el óptimo",
                "subtitle": f"{graph.subtitle}. Fijas en el óptimo: {held}",
            }
        )
    return graph, warnings


def _decision_names(req: LPRequest) -> list[str]:
    if req.variable_names:
        return list(req.variable_names)
    names: set[str] = set(req.objective)
    for c in req.constraints:
        names.update(c.coeffs)
    return sorted(names)


def _free_names(req: LPRequest, names: list[str]) -> tuple[list[str], list[str]]:
    requested = _dedupe(list(req.graph_variables or []))
    if not requested:
        if len(names) <= 3:
            return list(names), []
        return list(names[:3]), []
    known = [n for n in requested if n in names]
    unknown = [n for n in requested if n not in names]
    if len(known) >= 2:
        return known[:3], unknown
    if len(names) <= 3:
        return list(names), unknown
    return list(names[:3]), unknown


def _dedupe(names: list[str]) -> list[str]:
    seen: set[str] = set()
    out: list[str] = []
    for name in names:
        if name in seen:
            continue
        seen.add(name)
        out.append(name)
    return out


def _reduce(
    req: LPRequest, x_star: dict[str, float], free: list[str]
) -> tuple[LPRequest, float, dict[str, float]]:
    names = _decision_names(req)
    fixed = {name: float(x_star.get(name, 0.0)) for name in names if name not in free}
    z_offset = sum(float(req.objective.get(name, 0.0)) * value for name, value in fixed.items())
    constraints: list[LPConstraint] = []
    for c in req.constraints:
        shift = sum(float(c.coeffs.get(name, 0.0)) * value for name, value in fixed.items())
        coeffs = {name: float(c.coeffs.get(name, 0.0)) for name in free}
        if all(abs(coeffs[name]) < EPS for name in free):
            continue
        constraints.append(
            LPConstraint(id=c.id, coeffs=coeffs, sense=c.sense, rhs=_clean(float(c.rhs) - shift))
        )
    bounds = None
    if req.bounds:
        bounds = {name: req.bounds[name] for name in free if name in req.bounds}
    reduced = LPRequest(
        sense=req.sense,
        objective={name: float(req.objective.get(name, 0.0)) for name in free},
        constraints=constraints,
        variable_names=list(free),
        bounds=bounds,
        include_iterations=False,
        include_sensitivity=False,
        include_graph=True,
    )
    return reduced, z_offset, fixed

File: worker/src/modules/lp/graphNd.ts
import type { GraphXY } from "../../schema";
import { build2dGraph } from "./graph2d";
import { build3dGraph } from "./graph3d";
import type { LPConstraint, LPRequest } from "./types";
import { collectVarNames } from "./types";

const EPS = 1e-9;

export function buildLpGraph(
  req: LPRequest,
  xStar: Record<string, number>,
  zStar: number,
): { graph: GraphXY | null; warnings: string[] } {
  const names = collectVarNames(req);
  const warnings: string[] = [];
  if (names.length < 2) return { graph: null, warnings };
  const picked = freeNames(req, names);
  warnings.push(...picked.warnings);
  if (picked.free.length < 2) return { graph: null, warnings };
  const sameSet = picked.free.length === names.length && names.every((name) => picked.free.includes(name));
  if (sameSet && picked.free.length === 2) return { graph: build2dGraph(req, xStar, zStar), warnings };

  const reduced = reduce(req, names, xStar, picked.free);
  const zFree = zStar - reduced.zOffset;
  let graph: GraphXY | null;
  if (picked.free.length === 2) graph = build2dGraph(reduced.req, xStar, zFree, reduced.zOffset);
  else {
    graph = build3dGraph(reduced.req, xStar, zFree, reduced.zOffset, sameSet ? null : reduced.fixed);
    if (!graph && !sameSet) {
      warnings.push("No se pudo construir el poliedro 3D. Elige otras tres variables o un corte de dos.");
    }
  }
  if (!graph) return { graph: null, warnings };
  if (!sameSet && picked.free.length === 2) {
    const held = Object.entries(reduced.fixed)
      .map(([name, value]) => `${name} = ${fmtNum(value)}`)
      .join(", ");
    graph = {
      ...graph,
      title: "Corte por el óptimo",
      subtitle: `${graph.subtitle ?? ""}. Fijas en el óptimo: ${held}`,
    };
  }
  return { graph, warnings };
}

function freeNames(req: LPRequest, names: string[]): { free: string[]; warnings: string[] } {
  const requested = dedupe(req.graph_variables ?? []);
  if (!requested.length) {
    return { free: names.length <= 3 ? [...names] : names.slice(0, 3), warnings: [] };
  }
  const known = requested.filter((name) => names.includes(name));
  const unknown = requested.filter((name) => !names.includes(name));
  const warnings = unknown.length
    ? [`Variables de gráfico ignoradas porque no están en el modelo: ${unknown.join(", ")}`]
    : [];
  if (known.length >= 2) return { free: known.slice(0, 3), warnings };
  return { free: names.length <= 3 ? [...names] : names.slice(0, 3), warnings };
}

function dedupe(names: string[]): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  for (const name of names) {
    if (seen.has(name)) continue;
    seen.add(name);
    out.push(name);
  }
  return out;
}

function reduce(req: LPRequest, names: string[], xStar: Record<string, number>, free: string[]) {
  const fixed: Record<string, number> = {};
  for (const name of names) {
    if (!free.includes(name)) fixed[name] = xStar[name] ?? 0;
  }
  const zOffset = Object.entries(fixed).reduce((sum, [name, value]) => sum + (req.objective[name] ?? 0) * value, 0);
  const constraints: LPConstraint[] = [];
  for (const c of req.constraints) {
    const shift = Object.entries(fixed).reduce((sum, [name, value]) => sum + (c.coeffs[name] ?? 0) * value, 0);
    const coeffs: Record<string, number> = {};
    for (const name of free) coeffs[name] = c.coeffs[name] ?? 0;
    if (free.every((name) => Math.abs(coeffs[name]) < EPS)) continue;
    constraints.push({ id: c.id, coeffs, sense: c.sense, rhs: clean(c.rhs - shift) });
  }
  const bounds = req.bounds
    ? Object.fromEntries(free.filter((name) => name in req.bounds!).map((name) => [name, req.bounds![name]]))
    : null;
  const next: LPRequest = {
    sense: req.sense,
    objective: Object.fromEntries(free.map((name) => [name, req.objective[name] ?? 0])),
    constraints,
    variable_names: [...free],
    bounds,
    include_iterations: false,
    include_sensitivity: false,
    include_graph: true,
  };
  return { req: next, zOffset, fixed };
}

function fmtNum(value: number): string {
  if (Math.abs(value - Math.round(value)) < 1e-8) return String(Math.round(value));
  return value.toPrecision(4).replace(/\.?0+$/, "");
}

function clean(value: number): number {
  const rounded = Number(value.toFixed(10));
  return Math.abs(rounded) < 1e-12 ? 0 : rounded;
}
